src/Illuminate/Database/Eloquent/Concerns/QueriesRelationships.php

```diff
@@ -556,16 +557,26 @@ public function withAggregate($relations, $column, $function = null)
                 $expression = $column;
             }
 
-            // Here, we will grab the relationship sub-query and prepare to add it to the main query
-            // as a sub-select. First, we'll get the "has" query and use that to get the relation
-            // sub-query. We'll format this relationship name and append this column if needed.
-            $query = $relation->getRelationExistenceQuery(
-                $relation->getRelated()->newQuery(), $this, new Expression($expression)
-            )->setBindings([], 'select');
+            if ($preserveRelationState) {
+                $query = $relation->getQuery()->whereColumn(
+                    $relation->getQualifiedParentKeyName(), '=', $relation->getExistenceCompareKey()
+                );
+            } else {
+                // Here, we will grab the relationship sub-query and prepare to add it to the main query
+                // as a sub-select. First, we'll get the "has" query and use that to get the relation
+                // sub-query. We'll format this relationship name and append this column if needed.
+                $query = $relation->getRelationExistenceQuery(
+                    $relation->getRelated()->newQuery(), $this, new Expression($expression)
+                )->setBindings([], 'select');
+            }
 
             $query->callScope($constraints);
 
-            $query = $query->mergeConstraintsFrom($relation->getQuery())->toBase();
+            if ($preserveRelationState) {
+                $query = $query->getConnection()->query()->from($query, $name.'_aggregate')->select(new Expression('count(*)'));
```

Comment:
`count(*)` is hardcoded here but this method also supports `exists`, `max`, `min`, `avg`, or `sum` as indicated by the `$function` argument.